src/stateManager/utils/FraudProofService.ts

```diff
@@ -0,0 +1,178 @@
+import Storage from "@/storage";
+import { Block, StateSnapshot } from "@/models";
+import { SignedBlockStruct } from "@typechain-types/contracts/V1/types/DataTypes";
+import {
+    BlockDoubleSignProofStruct,
+    BlockInvalidStateTransitionProofStruct,
+    InvalidTimestampProofStruct
+} from "@typechain-types/contracts/V1/types/FraudProofTypes";
+import { PreviousEntity } from "./channelValidation";
+
+const NULL = "0x00";
+
+// ────────────────────── CUSTOM EXCEPTIONS ─────────────────────
+
+/**
+ * Base class for validation fraud exceptions
+ */
+export abstract class ValidationFraudException extends Error {
+    constructor(message: string) {
+        super(message);
+        this.name = this.constructor.name;
+    }
+}
+
+export class DoubleSignException extends ValidationFraudException {
+    constructor(
+        public readonly block1: SignedBlockStruct,
+        public readonly block2: SignedBlockStruct
+    ) {
+        super("Double sign fraud detected");
+    }
+}
+
+export class InvalidStateTransitionException extends ValidationFraudException {
+    constructor(
+        public readonly previousEntity: PreviousEntity,
+        public readonly invalidBlock: SignedBlockStruct
+    ) {
+        super("Invalid state transition fraud detected");
+    }
+}
+
+export class InvalidTimestampException extends ValidationFraudException {
+    constructor(
+        public readonly previousEntity: PreviousEntity,
+        public readonly invalidBlock: SignedBlockStruct
+    ) {
+        super("Invalid timestamp fraud detected");
+    }
+}
+
+export class InvalidLeaderException extends ValidationFraudException {
+    constructor(
+        public readonly previousEntity: PreviousEntity,
+        public readonly invalidBlock: SignedBlockStruct
+    ) {
+        super("Invalid leader fraud detected");
+    }
+}
+
+// ────────────────────── FRAUD PROOF SERVICE ─────────────────────
+
+/**
+ * Service class for handling fraud proof creation and validation
+ */
+export class FraudProofService {
+    constructor(private readonly storage: Storage) {}
+
+    /**
+     * Create invalid state transition proof
+     */
+    createInvalidStateTransitionProof(
```

Comment:
so instead of throwing the errors which have the same arguments as these functions, just call these functions on the spot and all they do is `create` the `specific fraud proof`, `encode` it and `wrap it` into a `general fraud proof struct` as defined in solidity and perisit it in storage in a way where it's easy to retrive fraud proofs based for participants (e.g. `getFraudproof(participant)`) - so everything is in storage and when we eventually call `dispute()` later - we don't have to pass fraud proofs to it. It can fetch it from storage in O(1) for every participant.